HIP/hip-904-frictionless-airdrop.md

```diff
@@ -0,0 +1,711 @@
+
+However, it is important to have a separate HAPI transaction type for a few reasons
+
+- the ability to call out the potential separate behaviours based on maxAutomaticAssociation, whereby transfer can end up in a pending transfer state or a consumption of an association slot.
+- cleaner conceptual understanding
+- naming allows the grouping of functionality with Cancel and Claim which complete the pending state a TokenAirdrop may creates
+- explicit specification in record stream.
+- improved developer experience
+
+Similar restrictions around transfer list and aggregation across sender and receiver from CryptoTransfer will be applied to TokenAirdop e.g. max 10 transfers per transaction
+
+At the time of this HIP `TokenAirdropTransaction` will not be supported by the Scheduled Transaction Service. A future HIP may explore this support.
+
+#### Cancel
+
+```protobuf
+message TokenPendingAirdrop {
+    /**
+    * Account that issue the airdrop trasnaction
+    */
+	AccountID sender_id = 1;
+
+	/**
+    * Account that was the intended reciever of the airdrop
+    */
+    AccountID receiver_id = 2;
+
+	/**
+    * The unique token identifier. If it's a FT then it's just the TokenID.
+	* If it's an NFT then NftID which captures both the TokenID and serial id
+    */
+	oneof token_id {
+		NftID non_fungible_token = 3;
+		TokenID fungible_token = 4;
+	}	
+}
+
+```
+
+```protobuf
+/**
+ * A token cancel will allow an airdrop sender to revert clear a pending transfer from state.
+ * Afer this no pending transfer will exist for a potential receiver to claim.  
+ * This transfer will be represented in the transaction body and will not need restating in the record file.
+ * All cancellations must succeed for the transaction to succeed.
+ */
+message TokenCancelAirdropTransactionBody {
+    /**
+    * A list of up to 10 pending airdrops that may be cancelled. If there are
+    * duplicate pending_airdrops, then the transaction will fail.
+    */
+    repeated TokenPendingAirdrop pending_airdrops = 1;
+}
+```
+
+#### Claim
+
+```protobuf
+/**
+ * A token cancel will clear a pending transfer from state. 
+ * This transfer will be represented in the transaction body and will not need restating in the record file.
+ * All claims must succeed for the transaction to succeed.
+ */
+message TokenClaimAirdropTransactionBody {
+    /**
+    * The pending airdrop to be accepted. The transaction must be signed by
+    * the `receiverAccountID`'s key as contained in the pending airdrop.
+    * A list of up to 10 pending airdrops that may be claimed. If there are
+    * duplicate pending_airdrops, then the transaction will fail.
+    */
+    repeated TokenPendingAirdrop pending_airdrops = 1;
```

Comment:
Same comment as cancel wrt efficiency (bytes) of the transaction payload.

Reply:
Here, I wonder if just accepting a TokenID or list of serial_numbers would be better?
The accountID is known (the submitter/payer); so all that's left is knowing which pending airdrops for that account, and TokenID will identify any fungible/common tokens, while TokenID plus serial_number will identify non-fungible/unique tokens.
Example:
```
    /**
     * The ID of a token with unclaimed airdrops pending for the
     * submitting account.<br/>
     * All pending airdrops to be claimed in this transaction SHALL be
     * airdrops of this token.
     */
    TokenID token_to_claim = 1;

    /**
    * A list of serial numbers of a non-fungible/unique token pending
    * for the submitting account that SHOULD be claimed.<br/>
    * This list MUST NOT have duplicate entries.<br/>
    * This list MUST be 10 items or less.<br/>
    * If the token to be claimed is a fungible/common token, then this
    * filed MUST be unset.
    */
    repeated int64 serial_numbers_to_claim = 2;
```

Reply:
AccountID is not always know because payer is not always the sender, so we have to explicitly capture it just like CryptoTransfer

Reply:
I see. That makes sense.  I had assumed a single airdrop would be one token and one sender; but looking at the message, it's an arbitrary mix thanks to using repeated transfer lists (each of which may be many individual transfers).

Reply:
Protobuf suggestions was re-reviewed with considerations for transaction I/O and state management and a balance was struck